src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,433 @@
+            // {
+            //     int y;
+            //     int x;
+            //     DoSomething(out x, out y);
+            // }
+            // void Callee(out int i, out int j) => DoSomething(out i, out j);
+            // 'y' might becomes the first declaration if using dictionary instead of array.
+            var parametersWithVariableDeclarationArgument = operationsWithVariableDeclarationArgument
+                .Select(argument => (
+                    argument.Parameter,
+                    semanticModel.GetSymbolInfo(argument.Children.Single().Syntax, cancellationToken).GetAnySymbol()?.Name))
+                .Where(parameterAndArgumentName => parameterAndArgumentName.Name != null)
+                .ToImmutableArray();
+
+            return new MethodParametersInfo(
+                parametersWithVariableDeclarationArgument,
+                parametersToGenerateFreshVariablesFor,
+                parameterToReplaceMap);
+        }
+
+        /// <summary>
+        /// Check if the parameter is referenced only once, and it is referenced as 'read'.
+        /// Determine a special case for a parameter that should be replaced by the argument instead of generating a declaration
+        /// for it.
+        /// Example:
+        /// Before:
+        /// void Caller(bool x)
+        /// {
+        ///     Callee(Foo(), Bar())
+        /// }
+        /// void Callee(int a, int b)
+        /// {
+        ///     DoSomething(a, b);
+        /// }
+        /// After:
+        /// void Caller(bool x)
+        /// {
+        ///     DoSomething(Foo(), Bar());
+        /// }
+        /// void Callee(int a, int b)
+        /// {
+        ///     DoSomething(a, b);
+        /// }
+        /// Parameters 'a' and 'b' are used only once in the Callee, and their value are read not write.
+        /// For this case just use the argument to replace the parameter.
+        /// Note: This might cause a semantic change. In the previous example, if it is
+        /// void Caller(bool x)
+        /// {
+        ///     Callee(Foo(), Bar())
+        /// }
+        /// void Callee(int a, int b)
+        /// {
+        ///     DoSomething(b, a);
+        /// }
+        /// Then this operation will change the order of evaluation.
+        /// </summary>
+        private static async Task<ImmutableArray<IArgumentOperation>> GetArgumentsReadOnlyOnceAsync(
+            Document document,
+            ImmutableArray<IArgumentOperation> arguments,
+            TMethodDeclarationSyntax calleeMethodNode,
+            CancellationToken cancellationToken)
+        {
+            var builder = ArrayBuilder<IArgumentOperation>.GetInstance();
+            foreach (var argument in arguments)
+            {
+                var parameterSymbol = argument.Parameter;
+                var allReferences = await SymbolFinder
+                    .FindReferencesAsync(parameterSymbol, document.Project.Solution, cancellationToken).ConfigureAwait(false);
```

Comment:
you can vastly speed this up also by only passing the document the method is declared in.